docs/man/mkrlconf.8
.\" Copyright 2015-2024 Roderick W. Smith (rodsmith@rodsbooks.com)
.\" May be distributed under the GNU Free Documentation License version 1.3 or
any later version
.TH "MKRLCONF" "8" "0.14.0" "Roderick W. Smith" "rEFInd Manual"
.SH "NAME"
mkrlconf \- Create a Linux kernel configuration file for rEFInd
.SH "SYNOPSIS"
.BI "mkrlconf "
[ \-\-force ]

.SH "DESCRIPTION"

To boot a Linux kernel directly, rEFInd must normally pass system-specific
parameters to help the kernel locate its initial RAM disk (initrd) file,
the installation's root filesystem, and so on. rEFInd stores this
information in a file called \fIrefind_linux.conf\fR, which is stored in
the same directory as the kernel(s) to which it applies. The \fImkrlconf\fR
script creates this configuration file in \fI/boot\fR, using the current
boot options (from \fI/proc/cmdline\fR) to populate
\fI/boot/refind_linux.conf\fR with boot options that are probably (but not
certainly) correct.

The file created in this way has three lines, which correspond to three
entries on the rEFInd suboptions menu. The first entry boots using the
options found in \fI/proc/cmdline\fR. The second entry boots using the same
options as the first, but with \fBsingle\fR added. The third entry boots
with minimal options of \fBro root={CURRENT_ROOT_DEVICE}\fR, where
\fB{CURRENT_ROOT_DEVICE}\fR identifies the current root (\fI/\fR)
filesystem. Users may manually edit the \fIrefind_linux.conf\fR file to
suit their needs, of course.

.SH "OPTIONS"

.TP
.B \-\-force
Ordinarily, if \fImkrlconf\fR finds an existing
\fI/boot/refind_linux.conf\fR file, it refuses to replace it. The
\fB\-\-force\fR option causes \fImkrlconf\fR to replace the existing file
in favor of one it generates.

.SH "AUTHORS"
Primary author: Roderick W. Smith (rodsmith@rodsbooks.com)

.SH "SEE ALSO"
.BR mvrefind (8),
.BR refind-install (8),
.BR refind-sb-healthcheck (8).

\fIhttps://www.rodsbooks.com/refind/\fR

.SH "AVAILABILITY"
The \fBmkrlconf\fR command is part of the \fIrEFInd\fR package and is
available from Roderick W. Smith.


NEWS.txt
0.14.2 (?/??/2024):
-------------------

- Fixed bugs in RPM & Debian packaging related to Secure Boot:
  - The post-install script would sometimes fail to detect that Secure
    Boot was enabled, thus creating a non-bootable installation.
  - The post-install script assumed that all Secure Boot installations
    used Shim or PreLoader, thus failing to correctly install without Shim
    on systems that used custom db keys.

- Removed code in RPM & Debian packaging scripts that removed an existing
  rEFInd efibootmgr entry. This is obviated by code added to refind-install
  some time ago.

- Updated Debian packaging to require debhelper >= 11 rather than 9, since
  the latter is deprecated.

- Restructured code to detect most disk-based tools. The result is simpler
  code, and may slightly alter what tools are found, likely increasing the
  selection on systems with multiple disks or when tool binaries are stored
  in distribution directories.

- Removed EFI\tools\memtest86, EFI\tools\memtest, EFI\memtest86, and
  EFI\memtest from directories to be scanned for tools. This was done for
  performance reasons.

- Added also_scan_tool_dirs token, for specifying additional directories to
  scan for tools.

- Added indication of whether rEFInd was booted via Shim ("via Shim") or not
  ("direct") to the Info/About screen, on the "Platform" line, after the
  Secure Boot "active" vs. "inactive" indication. Note that "via Shim" may
  be specified even if Secure Boot is inactive, if the system is configured
  to boot rEFInd via Shim but Secure Boot is disabled. This information was
  already written to the refind.log file, when log_level is set to 1 or
  above.

0.14.1 (3/24/2024):
-------------------

- Fixed bug that caused JPEG decoding to produce garbled images when rEFInd
  was compiled with most Linux toolchains. It happened to work fine when
  compiled with Ubuntu (which I use) or Gentoo, but not with Arch, Debian,
  Fedora, or openSUSE.

- Fixed a memory leak in JPEG handling code.

- Fixed a bug that could cause some EFI programs (notably Pete Batard's
  drivers on ARM64 systems) to be incorrectly marked as invalid.

- Improved error checking of log file creation and rotation. Bugs in
  Tow-Boot (aka Das U-Boot) 8225.40 on a Raspberry Pi were interacting with
  rEFInd's minimal error checking on log file rotation, causing rEFInd to
  crash when attempting to rotate the log file. This is now fixed.

- Excise most BIOS/CSM/legacy scanning and boot code from builds that are
  not on x86/IA32 or x86-64/AMD64/X64. This code is useless without a CSM,
  which is a component that is useful only on x86/x86-64 systems. Removing
  this useless code reduces binary size and minimizes the risk of
  BIOS/CSM/legacy-related bugs affecting ARM64/AARCH64 builds (and
  conceivably others, in the future).

- Added linux_prefixes token, which specifies string(s) that, when present
  at the start of a filename, denote a Linux kernel.

- Replaced "bzImage" with "Image" as a kernel identifier on ARM64 systems.
  (This is unchanged on x86/x86-64 systems.) Some Linux distributions call
  their Linux kernels Image, Image.gz, or something else that starts with
  "Image".

- Worked around a bug in at least one UEFI that could cause rEFInd to freeze
  when doing a self-install.

- Added code to Btrfs driver to properly detect RAID1C3/4 configurations.
  Code was borrowed from GRUB 2 by Ilya Guterman.

- Added icon for EndeavourOS, as submitted by Jack Rubacha.

- Fixed bug in refind-install that caused --ownhfs option to fail.

- If Memtest is discovered in a boot loader location, move its entry to the
  tools line, rather than keep it in the boot loader line, if "memtest" is
  among the "showtools" options. This helps de-clutter the boot loader line
  with recent versions of Ubuntu. (For some reason, though, it fails to
  launch from some filesystems, with a "not found" error, in either menu
  position; but it launches fine if copied to the ESP.)

- Added memtest86+x64.efi and memtest86+ia32.efi (for AMD64 and IA32
  architectures, respectively) as valid memtest filenames; and removed
  bootx64.efi (and related architecture-specific names) as valid names,
  since the latter conflict with the new option to recognize memtest in a
  boot loader directory.

- Remove duplicate directory names when scanning for tools. Duplicates can
  happen because of misconfiguration or because of the new code to find
  Memtest in regular boot loader locations, if they happen to be in matching
  directories on different disks (as in multiple Linux distributions on one
  computer, with Memtest in the /boot directory on each one).

- Updated LodePNG to the latest (version 20230410).

0.14.0 (3/4/2023):
------------------

- Added ability to launch gzip-compressed loaders. This is useful mainly on
  ARM64 platforms.

- Added support for .sbat sections in rEFInd binaries. These are needed
  for recent versions of Shim to launch follow-on programs, including
  rEFInd.

- Added new refind-sb-healthcheck script, which helps manage Secure Boot
  keys. See the docs/refind/secureboot.html documentation and
  docs/man/refind-sb-healthcheck.8 man page for details.

- Added new follow_symlinks option. This causes rEFInd to read loaders that
  are symbolic links on filesystems that support this feature. In most
  cases, these symlinks are redundant; but in some cases (such as on
  openSUSE) following the symlinks is necessary to read a kernel. Credit to
  dakanji for adding this feature.

- Minor cleanup of mkrlconf and refind-install scripts; should not affect
  functionality

- Adrien Destugues submitted a new icon for the Haiku OS, based on an
  official Haiku OS icon.

- The ext4fs driver now ignores the metadata_csum_seed flag, which is being
  set by default in e2fsprogs 1.47.0 and later. Thanks to Martin Whitaker
  for the relevant patch.

- Added several new Secure Boot keys for Red Hat in the keys directory.

mkrlconf.sh
#!/bin/bash

# Script to create a refind_linux.conf file for the current Linux
# installation.

# copyright (c) 2012-2015 by Roderick W. Smith
#
# This program is licensed under the terms of the GNU GPL, version 3,
# a copy of which should be distributed with this program.

# Usage:
#
# ./mkrlconf.sh [--force]
#
# Options:
#
#   --force  -- Overwrite an existing file (default is to not replace existing file)

# Revision history:
#
#  0.9.0 -- Added check for OS type, to keep from running pointlessly on OS X
#  0.7.7 -- Fixed bug that caused stray PARTUUID= and line breaks in generated file
#  0.5.1 -- Initial release
#
# Note: mkrlconf.sh version numbers match those of the rEFInd package
# with which they first appeared.

RLConfFile="/boot/refind_linux.conf"

if [[ `uname -s` != "Linux" ]] ; then
   echo "This script is intended to be run from Linux. Aborting!"
   echo ""
   exit 1
fi

if [[ ! -f $RLConfFile || $1 == "--force" ]] ; then
   if [[ -f /etc/default/grub ]] ; then
      # We want the default options used by the distribution, stored here....
      source /etc/default/grub
   fi
   RootFS=`df / | grep dev | cut -f 1 -d " "`
   StartOfDevname=`echo $RootFS | cut -b 1-7`
   if [[ $StartOfDevname == "/dev/sd" || $StartOfDevName == "/dev/hd" ]] ; then
      # Identify root filesystem by UUID rather than by device node, if possible
      Uuid=`blkid -o export -s UUID $RootFS 2> /dev/null | grep UUID=`
      if [[ -n $Uuid ]] ; then
         RootFS=$Uuid
      fi
   fi
   DefaultOptions="$GRUB_CMDLINE_LINUX $GRUB_CMDLINE_LINUX_DEFAULT"
   echo "\"Boot with standard options\"        \"ro root=$RootFS $DefaultOptions \"" > $RLConfFile
   echo "\"Boot to single-user mode\"          \"ro root=$RootFS $DefaultOptions single\"" >> $RLConfFile
   echo "\"Boot with minimal options\"         \"ro root=$RootFS\"" >> $RLConfFile
else
   echo "Existing $RLConfFile found! Not overwriting!"
   echo "To force overwriting, pass the --force option."
   echo ""
fi
